reduce unnecessary info messages

naiming-zededa opened the pull request:
Signed-off-by: Naiming Shen <naiming@zededa.com>
Some of the info message updates do not reflect changes on real status of the device
they mainly due to new timestamp changes from some structure status check.
- this patch is to find the changes due to insignificant changes and suppress the upload
- but add a periodical publish of info about 10 mins to advertise the more recent timestamps

eriknordmark commented on pkg/pillar/cmd/zedagent/zedagent.go:
```diff
@@ -1791,7 +1801,18 @@ func handleDPCLImpl(ctxArg interface{}, key string,
+
+	status := statusArg.(types.DevicePortConfigList)
+	var exclude []string
+	exclude = append(exclude, "LastFailed:")
```
Don't we want to see changes to LastFailed?

naiming-zededa replied:
I was thinking when first time it failed, there is some 'real change' in the struct, so we'll publish right away. but later on it always fail the same way, only the timestamp change, we don't care. but the 10 min periodical publish will reflect the 'LastFailed' correctly since it gets the most recent one.

eriknordmark replied:
ok

eriknordmark commented on pkg/pillar/cmd/zedagent/zedagent.go:
```diff
@@ -1972,6 +1993,7 @@ func handleZbootStatusImpl(ctxArg interface{}, key string,
 	}
 	log.Functionf("handleZbootStatusImpl: for %s", key)
 	// nothing to do
+	log.Errorf("in zboot impl, hasRealChange")
```
Why error?

naiming-zededa replied:
Forgot to move this to Functionf. will change

eriknordmark commented on pkg/pillar/cmd/zedagent/zedagent.go:
```diff
@@ -2042,6 +2072,28 @@ func handleNodeAgentStatusDelete(ctxArg interface{}, key string,
+	diffs := strings.Split(diffStr, "\n")
+	for _, diff := range diffs {
+		var isExcluded bool
+		if strings.HasPrefix(diff, "-") || strings.HasPrefix(diff, "+") {
+			for _, exclude := range excludeChgStr {
+				if strings.Contains(diff, exclude) {
+					isExcluded = true
+					break
```
Is there anything in the cmp package which has the ability to ignore certain fields? 
It would be nicer to not depend on the string output format of cmp.Diff.

naiming-zededa replied:
from my investigation, it does not appear so. they have something like if the floating point number change is smaller than a value, then let diff pass kind of things. and the structure also can be hierarchical as in our case.

eriknordmark replied:
@naiming-zededa is there a better way?

eriknordmark reviewed (commented):
We should talk about splitting out parts of the device info proto message into a hardware info message (about things which never change or only change when some new disk, network, memory card is plugged in). That will make the device info message a lot smaller.

naiming-zededa commented:
> We should talk about splitting out parts of the device info proto message into a hardware info message (about things which never change or only change when some new disk, network, memory card is plugged in). That will make the device info message a lot smaller.

Makes sense.

naiming-zededa commented:
changed to using copy & clear then compare.

giggsoff commented on pkg/pillar/cmd/zedagent/zedagent.go:
```diff
@@ -1805,6 +1840,26 @@ func handleDPCLDelete(ctxArg interface{}, key string, statusArg interface{}) {
+func dpcHasRealChange(dpcList types.DevicePortConfigList) bool {
+	dummy := dpcList
+	for i, l := range dummy.PortConfigList {
+		l.LastFailed = time.Time{}
+		l.LastSucceeded = time.Time{}
+		l.LastError = ""
```
Shouldn't we check for updates of error here (and in another places) and send info in case of any?

eriknordmark commented on pkg/pillar/cmd/zedagent/zedagent.go:
```diff
@@ -1805,6 +1840,26 @@ func handleDPCLDelete(ctxArg interface{}, key string, statusArg interface{}) {
 	triggerPublishDevInfo(ctx)
 }
 
+func dpcHasRealChange(dpcList types.DevicePortConfigList) bool {
```
Would be good to add a comment saying it clears the fields which we want to ignore in the comparison.

naiming-zededa replied:
ok.

eriknordmark commented on pkg/pillar/cmd/zedagent/zedagent.go:
```diff
@@ -80,6 +81,11 @@ var diagMetrics types.MetricsMap
 var zrouterMetrics types.MetricsMap
 
+// modified for real change detection, some fields zeroed out
+var modDPC types.DevicePortConfigList
+var modNodeStatus types.NodeAgentStatus
+var modDNS types.DeviceNetworkStatus
```
These shouldn't be global variables. When comparing A and B you can clear the dont-care fields in both A and B and then do the compare.

eriknordmark reviewed (changes requested):
Please remove the global variables used in the comparison

naiming-zededa commented:
modified to have the clear & compare A & B, DNS already has the old data, added two new into contexts for DNCL and NodeAgent which were not saved before. Add back the "LastError" as @giggsoff commented on.

eriknordmark reviewed (approved):
Kick off tests

eriknordmark reviewed (approved):
Run eden again

eriknordmark reviewed (commented):
Can you squash the two commits? I don't think they make sense as patched in isolation.

naiming-zededa commented:
will do the squash

naiming-zededa commented:
let me rename a few things.

rvs reviewed (commented):
LGTM
